codec-http/src/main/java/io/netty/handler/codec/http/multipart/MemoryFileUpload.java

```diff
@@ -132,35 +132,37 @@ public String toString() {
 
     @Override
     public FileUpload copy() {
-        MemoryFileUpload upload = new MemoryFileUpload(getName(), getFilename(), getContentType(),
-                getContentTransferEncoding(), getCharset(), size);
-        ByteBuf buf = content();
-        if (buf != null) {
-            try {
-                upload.setContent(buf.copy());
-                return upload;
-            } catch (IOException e) {
-                throw new ChannelException(e);
-            }
-        }
-        return upload;
+        final ByteBuf content = content();
+        return replace(content != null ? content.copy() : content);
     }
 
     @Override
     public FileUpload duplicate() {
-        MemoryFileUpload upload = new MemoryFileUpload(getName(), getFilename(), getContentType(),
-                getContentTransferEncoding(), getCharset(), size);
-        ByteBuf buf = content();
-        if (buf != null) {
+        final ByteBuf content = content();
+        return replace(content != null ? content.duplicate() : content);
+    }
+
+    @Override
+    public FileUpload retainedDuplicate() {
+        final ByteBuf content = content();
+        return replace(content != null ? content.retainedDuplicate() : content);
```

Comment:
if a `ChannelException` is thrown we need to take care to release the `retained` buffer in this case